### PR TITLE
add support for returning aux data from grad

### LINKED ISSUE #366: add a way to return side-information from grad
For example, a user may want to return a statistic or auxiliary data structure from the Python function to which `grad` is applied, not to be considered as part of the value of the mathematical function to be differentiated.

### LINKED ISSUE #483: Error when passing the JaxTuple as side-information of the loss function
An issue related to #366, which is trying to pass the side-information out from grad.

The code to reproduce the error is pasted as follows.
```
from __future__ import print_function, division
import jax.numpy as np
from jax import grad, jit, vmap, vjp
from jax import random
from jax.tree_util import tree_map
from jax import core

from jax.experimental import optimizers

key = random.PRNGKey(0)

w = np.array([0., 1.])
x = random.normal(key, shape=(2,))
lr = np.array(1.)

opt_init, opt_update = optimizers.adam(1e-4)
opt_state = opt_init(w)

dw = grad(lambda w: np.square(np.dot(w, x) - 1.))
          
def loss(lr, opt_state):
  w = optimizers.get_params(opt_state)
  dummy_grad = dw(w) * lr
  opt_state = opt_update(0, dummy_grad, opt_state)
  w = optimizers.get_params(opt_state)
  return np.sum(np.square(w)), opt_state

# A trick from https://github.com/google/jax/issues/366
@jit
def mygrad(opt_state, lr):
  (l, extra), pullback = vjp(loss, lr, opt_state)
  all_zeros = tree_map(lambda x: core.pack(map(np.zeros_like, x)), opt_state)
  seeds = (np.asarray(1.0, np.float32), all_zeros)
  dx,_ = pullback(seeds)
  return dx, extra

grad, opt_state = mygrad(opt_state, lr)
```

Also the colab with the same code, if it's helpful https://colab.research.google.com/drive/1azubGbEk4ZS0EEVnm9iMyIwheDEN210m

### PR DESCRIPTION
**Pull request by mattjj**
fixes #366 and fixes #483 

The auxiliary data can be any pytree. Here's an example:

```python
from jax import grad

def f(x):
  val = x**2
  aux = [x**3, x**4]
  return val, aux

g, aux = grad(f, has_aux=True)(2.)
print(g)  # 4. == grad(lambda x: x**2)(2.)
print(aux)  # [8., 16.]
```

Thoughts on that API? An alternative would be to add more functions, like `grad_and_aux`, but then we might also need `val_and_grad_and_aux`. Ugh.

Also, suggestions welcome on how to tighten up the code.

### CODE REVIEW

File: jax/api.py [hunk 1]
```diff
@@ -179,13 +179,17 @@ def grad(fun, argnums=0):
       arrays with shape `(1,)` etc.)
     argnums: Optional, integer or tuple of integers. Specifies which positional
       argument(s) to differentiate with respect to (default 0).
+    has_aux: Optional, bool. Indicates whether `fun` returns a pair where the
```
Comment (hawkinsp) [inline jax/api.py]: Another API option might be to have some sort of opaque box around return values that shouldn't be considered by the gradient computation (`StopGradientBox`? Not a serious name suggestion.)

**Review by hawkinsp — APPROVED** [review]

File: jax/api.py (hunk 1 above)
Comment (mattjj) [reply jax/api.py]: The trouble with that is the aux values computed may depend on the input with respect to which we're differentiating, and so they're already "infected" with tangents. We'd like to remove that infection for the tangent values associated with this particular differentiation trace, but not go full `lax.stop_gradient` because we actually want to preserve tangents associated with outer tracers that might want to differentiate through what we consider to be auxiliary data! The tests cover cases like that one.

Comment (mattjj) [reply jax/api.py]: I thought a bit more and I understand better what you're saying: this is an API choice, not an implementation choice. That's interesting... though it might be a bit heavyweight since we'd need to register those as valid JaxTypes. I think that cost is high enough that it's probably not worthwhile.
